// Copyright 2010 Google Inc. All Rights Reserved.
// Author: ptab@google.com (Piotr Tabor)
// Enums for internal usage.


package supersonic;

// Lists of functions used to realize expressions.
//
// If the tag % 4 is:
// - 0, then the operator is general, and has no policy-specific versions.
// - 1, then the operator is quiet,
// - 2, then the operator is nulling,
// - 3, then the operator is signaling.
enum OperatorId {
  // Operators
  OPERATOR_ADD = 0;
  OPERATOR_MULTIPLY = 4;
  OPERATOR_SUBTRACT = 8;
  OPERATOR_DIVIDE_QUIET = 13;
  OPERATOR_DIVIDE_NULLING = 14;
  OPERATOR_DIVIDE_SIGNALING = 15;

  OPERATOR_CPP_DIVIDE_QUIET = 17;
  OPERATOR_CPP_DIVIDE_NULLING = 18;
  OPERATOR_CPP_DIVIDE_SIGNALING = 19;

  OPERATOR_MODULUS_NULLING = 26;
  OPERATOR_MODULUS_SIGNALING = 27;

  OPERATOR_NEGATE = 36;

  // Logical operations:
  OPERATOR_AND = 40;
  OPERATOR_OR = 44;
  OPERATOR_AND_NOT = 48;
  OPERATOR_NOT = 52;
  OPERATOR_XOR = 56;

  // Bitwise operations:
  OPERATOR_BITWISE_AND = 60;
  OPERATOR_BITWISE_OR = 64;
  OPERATOR_BITWISE_NOT = 68;
  OPERATOR_BITWISE_XOR = 72;
  OPERATOR_SHIFT_LEFT =  76;
  OPERATOR_SHIFT_RIGHT = 80;
  OPERATOR_BITWISE_ANDNOT = 84;

  OPERATOR_EQUAL = 100;
  OPERATOR_NOT_EQUAL = 104;
  // We implement GREATER and GREATER_OR_EQUAL in terms of LESS and
  // LESS_OR_EQUAL, respectively, to reduce the binary size.
  OPERATOR_LESS = 116;
  OPERATOR_LESS_OR_EQUAL = 120;

  // Verifiers (numeric -> bool):
  OPERATOR_IS_ODD = 140;
  OPERATOR_IS_EVEN = 144;
  OPERATOR_IS_FINITE = 148;
  OPERATOR_IS_INF = 152;
  OPERATOR_IS_NAN = 156;
  OPERATOR_IS_NORMAL = 160;

  // Control instructions:
  OPERATOR_CASE = 200;
  OPERATOR_IF = 204;
  OPERATOR_IN = 208;
  OPERATOR_IF_NULL = 220;
  OPERATOR_IS_NULL = 224;

  // Exact copy of orginal value.
  OPERATOR_COPY = 260;

  OPERATOR_CAST_QUIET   = 265;
  OPERATOR_CAST_NULLING = 266;
  OPERATOR_CAST_SIGNALING = 267;

  OPERATOR_BOOL_TO_NUMERIC = 268;
  OPERATOR_NUMERIC_TO_BOOL = 272;

  // All functions compiled into evaluator (so called built-in) go here.
  // Math functions.
  OPERATOR_ROUND = 300;
  OPERATOR_TRUNC = 304;
  OPERATOR_CEIL_TO_INT = 308;
  OPERATOR_FLOOR_TO_INT = 312;
  OPERATOR_ROUND_TO_INT = 316;

  OPERATOR_EXP = 320;
  OPERATOR_LN_QUIET = 325;
  OPERATOR_LN_NULLING = 326;
  OPERATOR_LN_SIGNALING = 327;
  OPERATOR_LOG10_QUIET = 329;
  OPERATOR_LOG10_NULLING = 330;
  OPERATOR_LOG10_SIGNALING = 331;
  OPERATOR_SQRT_QUIET = 333;
  OPERATOR_SQRT_NULLING = 334;
  OPERATOR_SQRT_SIGNALING = 335;
  OPERATOR_LOG_QUIET = 337;
  OPERATOR_LOG_NULLING = 338;
  OPERATOR_LOG_SIGNALING = 339;
  OPERATOR_CEIL = 342;
  OPERATOR_FLOOR = 346;
  OPERATOR_POW_QUIET = 353;
  OPERATOR_POW_NULLING = 354;
  OPERATOR_POW_SIGNALING = 355;
  OPERATOR_LOG2_QUIET = 357;
  OPERATOR_LOG2_NULLING = 358;
  OPERATOR_LOG2_SIGNALING = 359;
  OPERATOR_ABS = 360;
  OPERATOR_ROUND_WITH_MULTIPLIER = 364;

  // String functions.
  OPERATOR_LENGTH = 400;
  OPERATOR_LTRIM = 404;
  OPERATOR_RTRIM = 408;
  OPERATOR_TRIM = 412;
  OPERATOR_TOUPPER = 416;
  OPERATOR_TOLOWER = 420;

  OPERATOR_SUBSTRING_SIGNALING = 427;

  OPERATOR_CONCATENATE = 428;
  OPERATOR_TOSTRING = 432;
  OPERATOR_PARSE_STRING_QUIET = 437;
  OPERATOR_PARSE_STRING_NULLING = 438;

  OPERATOR_REGEXP_PARTIAL = 460;
  OPERATOR_REGEXP_FULL = 464;
  OPERATOR_REGEXP_EXTRACT = 468;
  OPERATOR_REGEXP_REPLACE = 472;

  OPERATOR_STRING_OFFSET = 476;
  OPERATOR_STRING_REPLACE = 480;

  // Datetime functions:
  OPERATOR_UNIXTIMESTAMP = 500;
  OPERATOR_FROMUNIXTIME = 504;
  OPERATOR_MAKEDATE = 508;
  OPERATOR_YEAR = 512;
  OPERATOR_MONTH = 516;
  OPERATOR_DAY = 520;
  OPERATOR_WEEKDAY = 524;
  OPERATOR_YEARDAY = 528;
  OPERATOR_HOUR = 532;
  OPERATOR_MINUTE = 536;
  OPERATOR_SECOND = 540;
  OPERATOR_MICROSECOND = 544;
  OPERATOR_QUARTER = 548;
  OPERATOR_DATE_TO_DATETIME = 552;  // Internal, equivalent of quiet cast.

  OPERATOR_DATEFORMAT = 560;

  OPERATOR_ADD_DAYS = 568;
  OPERATOR_ADD_MONTHS = 576;
  OPERATOR_ADD_MINUTES = 586;
  // Various others
  OPERATOR_FORMAT_SIGNALING = 603;

  // More datetime functions...
  OPERATOR_YEAR_LOCAL = 604;
  OPERATOR_QUARTER_LOCAL = 608;
  OPERATOR_MONTH_LOCAL = 612;
  OPERATOR_DAY_LOCAL = 616;
  OPERATOR_WEEKDAY_LOCAL = 620;
  OPERATOR_YEARDAY_LOCAL = 624;
  OPERATOR_HOUR_LOCAL = 628;
  OPERATOR_MINUTE_LOCAL = 632;

  OPERATOR_DATEFORMAT_LOCAL = 644;

  // Hashing/fingerprinting.
  OPERATOR_FINGERPRINT = 700;
  OPERATOR_HASH = 704;

  // Trigonometry
  OPERATOR_SIN = 800;
  OPERATOR_COS = 804;
  OPERATOR_TAN = 808;
  OPERATOR_ASIN = 812;
  OPERATOR_ACOS = 816;
  OPERATOR_ATAN = 820;
  OPERATOR_ATAN2 = 824;
  OPERATOR_SINH = 828;
  OPERATOR_COSH = 832;
  OPERATOR_TANH = 836;
  OPERATOR_ASINH = 840;
  OPERATOR_ACOSH = 844;
  OPERATOR_ATANH = 848;

  // For test purposes only - used in ternary column computers testing.
  OPERATOR_TERNARY_NO_CHECK = 1002;
  OPERATOR_TERNARY_FULL_CHECK = 1003;
  OPERATOR_TERNARY_ALLOC = 1004;
}
